### ISSUE TITLE
Frequent Leadership Changes When Timelock Cluster Under Load

### ISSUE THREAD

**Issue by harrybiddle** [palantir/atlasdb#1680]
On a large internal stack we have observed that when under load, the timelock cluster may be switching leaders.

<img width="677" alt="screen shot 2017-03-06 at 11 32 42" src="https://cloud.githubusercontent.com/assets/5887227/23608304/abe3ab10-0260-11e7-8484-6521948513f7.png">

In the graph above, I think that each new colour represents a new leader taking over. Need to confirm this. This behaviour started when we turned on HTTP2 (the start of the graph hitting 100%).

Further investigation needed.

**Comment by schlosna** [comment]
probably worth exposing some metrics on how many times a node has become leader & election rate

**Comment by berler** [comment]
We are potentially seeing this issue internally again on a different stack. See PDS-50340

**Comment by harrybiddle** [comment]
We'll be using our large internal stack tomorrow to investigate this. Current thinking is that PDS-50340 is a different issue.

**Comment by jeremyk-91** [comment]
I think I've root caused this.

BLUF: **AtlasDB with HTTP/2 does not handle requests which block for more than OkHttp's stream selection timeout correctly. This is by default 30 seconds.**

Short term fix (disable automated connection retry at the OkHttpClient level for lock services) will be out shortly, I'll file an AtlasDB issue for a longer term fix (upgrade okhttp to 3.3).

Presumably we only saw this issue on [redacted] because it was the only stack we have which had sufficient request volume to actually have someone block on finishing some transaction for more than 30 seconds.

OkHttp silently retries requests which suffer from timeouts, *even if they are POSTs* (https://github.com/square/okhttp/issues/1043). As at OkHttp 3.3 this was fixed to be _connect_ timeouts (https://github.com/square/okhttp/pull/2479), but AtlasDB and its clients use OkHttp 2. Any lock request that blocked for more than 30 seconds was treated as failed and the client would have retried using a different connection. However, because we queue incoming requests, the old client's request would reach the server and be granted even though the client wouldn't be listening for it.

For context - if you're wondering _how_ I found this, I looked at the identities of the HTTP FramedConnections across multiple thread dumps over time, and managed to reproduce it with a local TimeLock server and a local misbehaving TimeLock client. This only occurs over HTTP/2.

```java
    public static void main(String[] args) throws Exception {
        RemoteLockService proxy = AtlasDbHttpClients.createProxy(
                Optional.of(
                        SslSocketFactories.createSslSocketFactory(
                                SslConfiguration.of(Paths.get("..","Client_Truststore"))
                        )
                ),
                "https://localhost:8080/test",
                RemoteLockService.class);

        List<Future<LockRefreshToken>> futures = Lists.newArrayList();
        LockRequest request = LockRequest.builder(ImmutableSortedMap.of(
                StringLockDescriptor.of("foo"), LockMode.WRITE))
                .build();
        for (int i = 0; i < TRIALS; i++) {
            int currentTrial = i;
            futures.add(EXECUTOR_SERVICE.submit(() -> proxy.lock(String.valueOf(currentTrial), request)));
        }
        Thread.sleep(35 * 1_000); // more than 30s
        for (int i = 0; i < TRIALS; i++) {
            LockRefreshToken token = futures.get(i).get();
            proxy.unlock(token);
        }
        System.out.println("All done!");
    }
```

With TimeLock request log output
![retry-deadlock 1](https://cloud.githubusercontent.com/assets/5955510/24167645/5ecd5a22-0e6f-11e7-8c4f-24671e8fef07.png)
 as you'd expect:

```
127.0.0.1 - - [21/Mar/2017:19:04:20 +0000] "POST /test/lock/lock/4 HTTP/2.0" 200 63 "-" "unknown-atlasdb (unknown)" 228
WARN  [2017-03-21 19:06:25,666] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=6pzdoo8xl9xx, client=LockClient{"4"}, createdAt=SimpleTimeDuration{1490123060666 milliseconds}, expiresIn=SimpleTimeDuration{-5000 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:06:25 +0000] "POST /test/lock/lock/3 HTTP/2.0" 200 65 "-" "unknown-atlasdb (unknown)" 125185
WARN  [2017-03-21 19:08:30,668] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=27o98f4u5laak, client=LockClient{"3"}, createdAt=SimpleTimeDuration{1490123185668 milliseconds}, expiresIn=SimpleTimeDuration{-5000 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:08:30 +0000] "POST /test/lock/lock/0 HTTP/2.0" 200 64 "-" "unknown-atlasdb (unknown)" 250185
WARN  [2017-03-21 19:10:35,670] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=13drncd9i770e, client=LockClient{"0"}, createdAt=SimpleTimeDuration{1490123310669 milliseconds}, expiresIn=SimpleTimeDuration{-5001 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:10:35 +0000] "POST /test/lock/lock/1 HTTP/2.0" 200 65 "-" "unknown-atlasdb (unknown)" 375187
WARN  [2017-03-21 19:12:40,672] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=2fra2gruojlpx, client=LockClient{"1"}, createdAt=SimpleTimeDuration{1490123435671 milliseconds}, expiresIn=SimpleTimeDuration{-5001 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:12:40 +0000] "POST /test/lock/lock/2 HTTP/2.0" 200 64 "-" "unknown-atlasdb (unknown)" 500186
WARN  [2017-03-21 19:14:45,673] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=12lk7lapzpz63, client=LockClient{"2"}, createdAt=SimpleTimeDuration{1490123560673 milliseconds}, expiresIn=SimpleTimeDuration{-5000 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:14:45 +0000] "POST /test/lock/lock/5 HTTP/2.0" 200 64 "-" "unknown-atlasdb (unknown)" 625189
WARN  [2017-03-21 19:16:50,675] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=159ldjldmmqrs, client=LockClient{"5"}, createdAt=SimpleTimeDuration{1490123685674 milliseconds}, expiresIn=SimpleTimeDuration{-5001 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:16:50 +0000] "POST /test/lock/lock/6 HTTP/2.0" 200 65 "-" "unknown-atlasdb (unknown)" 750189
WARN  [2017-03-21 19:18:55,676] com.palantir.lock.impl.LockServiceImpl: Lock token HeldLocksToken{tokenId=2tfa68z0peuzs, client=LockClient{"6"}, createdAt=SimpleTimeDuration{1490123810676 milliseconds}, expiresIn=SimpleTimeDuration{-5000 milliseconds}, lockCount=1, firstLock=LockDescriptor [foo]=WRITE, versionId=null} was not properly refreshed and is now being reaped.
127.0.0.1 - - [21/Mar/2017:19:18:55 +0000] "POST /test/lock/lock/6 HTTP/2.0" 200 65 "-" "unknown-atlasdb (unknown)" 844914
```

**Comment by jeremyk-91** [comment]
My mistake; the above addresses only half of the problem. Another part is that Dropwizard sets request timeouts to be 30 seconds by default, which are reset if a byte is written or read (which explains the 30 seconds) and sends a HTTP/2 cancel at that time. This is amplified by OkHttp spamming retries when the socket is closed.

Proposal (short term):
* Bump the socket timeout to 60 seconds, to limit the occurrence of this in the first place.
* Disable connection retry; we seek to instead fail gracefully, blocking a small number of threads rather than chomping down a massive number of threads.

Longer term solutions:
* Disable connection retry (this is required)
* Implement a ClientBlockingRemoteLockService that is able to maintain some kind of heartbeat with the remote lock service to keep its requests alive?

**Comment by jeremyk-91** [comment]
#1727 doesn't fix the issue, though it was a useful part of the solution.

From internal ticket:

> I still believe the root cause as outlined above is correct, though the fix did not address the issue. The lock service still retries, though I suspect for a different reason which I noticed when comparing the AtlasDB codepaths and the paths in my tester CLI:
> 
> - We have another layer of retrying on top of OkHttp: FailoverFeignTarget. Now this class resets the failure count upon hitting a node with a 503, which the non-leaders will quickly return.
> So previously we would have [try 4x on leader, all fail], [try on non-leader 1], [try on non-leader 2] ad infinitum; now we have [try 1x on leader, fail], [try on non-leader 1], [try on non-leader 2] ad infinitum.
> - I haven't validated this with the [redacted] data, though we could (looking at request logs).

#1727 addresses this second set of retries, and I have a better smoke test that verifies that if N requests are made, only up to N will reach the server (at least under normal circumstances).

**Comment by jeremyk-91** [comment]
The cause we're aware of for this was addressed in #1772 and #1808. I'm thus going to close this out now - in the event that this resurfaces we can reopen.
